Add container-level security context support for Kubernetes orchestrator

htahir1 opened the pull request:
Previously, only pod-level security context settings could be configured, and container-level settings were limited to the privileged flag. This caused issues in restricted environments that enforce container security policies like runAsNonRoot and allowPrivilegeEscalation.

This change adds a container_security_context field to KubernetesPodSettings that allows users to specify container-level security attributes:
- runAsUser
- runAsNonRoot
- allowPrivilegeEscalation
- and other V1SecurityContext attributes

The settings are applied to all containers in the pod and are properly merged with existing security context settings (like the privileged flag).

Fixes #4141

## Describe changes
I implemented/fixed _ to achieve _.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] I have added tests to cover my changes.
- [ ] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] **IMPORTANT**: I made sure that my changes are reflected properly in the following resources:
  - [ ] [ZenML Docs](https://docs.zenml.io)
  - [ ] Dashboard: Needs to be communicated to the frontend team.
  - [ ] Templates: Might need adjustments (that are not reflected in the template tests) in case of non-breaking changes and deprecations.
  - [ ] [Projects](https://github.com/zenml-io/zenml-projects): Depending on the version dependencies, different projects might get affected.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

claude committed:
Add container-level security context support for Kubernetes orchestrator

Previously, only pod-level security context settings could be configured,
and container-level settings were limited to the privileged flag. This
caused issues in restricted environments that enforce container security
policies like runAsNonRoot and allowPrivilegeEscalation.

This change adds a container_security_context field to KubernetesPodSettings
that allows users to specify container-level security attributes:
- runAsUser
- runAsNonRoot
- allowPrivilegeEscalation
- and other V1SecurityContext attributes

The settings are applied to all containers in the pod and are properly
merged with existing security context settings (like the privileged flag).

Fixes #4141

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/zenml-io/zenml?pullRequest=4142) <br/>All committers have signed the CLA.

schustmi commented on src/zenml/integrations/kubernetes/orchestrators/manifest_utils.py:
```diff
@@ -234,6 +234,34 @@ def add_pod_settings(
+            from zenml.integrations.kubernetes import serialization_utils
+
+            # Deserialize the security context settings
+            security_context_dict = (
```
This converts it to the model, only to later convert back to the dict?

claude committed:
Refactor container security context to avoid unnecessary conversions

Remove inefficient model-to-dict-to-model conversion cycle by using
the dict directly for merging. Also remove unnecessary comments.

claude committed:
Move serialization_utils import to top of file

No need for inline import since there are no circular dependencies.

claude committed:
Simplify container security context merging logic

Remove unnecessary hasattr check and else branch since
container.security_context is always a V1SecurityContext object
(created in build_pod_manifest). This makes the code much cleaner
and more straightforward.

claude committed:
Simplify security context assignment to match existing patterns

Use direct dict assignment like all other pod settings (affinity,
tolerations, resources, env, etc). The kubernetes client accepts both
dicts and model objects, so no need for manual deserialization.

Also removed unused serialization_utils import.

claude committed:
Add clarifying note about container vs pod-level security context

Explains why container_security_context is needed separately from
additional_pod_spec_args, since some admission policies specifically
check container-level security settings.

schustmi commented on src/zenml/integrations/kubernetes/orchestrators/manifest_utils.py:
```diff
@@ -234,6 +234,11 @@ def add_pod_settings(
                 container.env_from = settings.env_from
 
+        if settings.container_security_context:
+            existing_dict = container.security_context.to_dict()
```
Are you sure this is always set?

claude committed:
Add defensive check for container security context existence

While security_context should always be set in the current code flow,
this adds a defensive check to handle edge cases where it might be None.
If security_context exists, merge with new settings. Otherwise, set directly.
